File: c++/0008-string-to-integer-atoi.cc
#include <gtest/gtest.h>

#include <string_view>

namespace stringToIntegerAtoi {

int solve(const std::string_view& s) {
  int i = 0;
  int sign = 1;
  long result = 0;

  // discard any leading whitespace
  while (i < s.length() && std::isspace(s[i])) {
    i++;
  }

  // is this a sign character?
  if (i < s.length() && (s[i] == '-' || s[i] == '+')) {
    sign = (s[i] == '-') ? -1 : 1;
    i++;
  }

  // from now on, we're only interested in digits
  while (i < s.length() && std::isdigit(s[i])) {
    result = result * 10 + (s[i] - '0');

    // how to handle overflow/underflow cases? clamp, I guess.
    if (sign * result > std::numeric_limits<int>::max()) {
      return std::numeric_limits<int>::max();
    }
    if (sign * result < std::numeric_limits<int>::min()) {
      return std::numeric_limits<int>::min();
    }

    i++;
  }

  return sign * result;
}

}  // namespace stringToIntegerAtoi

TEST(StringToIntegerAtoi, TestCases) {
  EXPECT_EQ(stringToIntegerAtoi::solve("42"), 42);
  EXPECT_EQ(stringToIntegerAtoi::solve("   -042"), -42);
  EXPECT_EQ(stringToIntegerAtoi::solve("1337c0d3"), 1337);
  EXPECT_EQ(stringToIntegerAtoi::solve("0-1"), 0);
}

File: swift/Sources/LeetCode/0008StringToIntegerAtoi.swift
public func stringToIntegerAtoi(_ s: String) -> Int {
    let chars = Array(s)
    
    var i = 0
    var sign = 1
    var result  = 0

    // discard leading whitespace
    while i < chars.count, chars[i].isWhitespace {
        i += 1
    }
    
    // look for sign character
    if i < chars.count, chars[i] == "-" || chars[i] == "+" {
        sign = chars[i] == "-" ? -1 : 1
        i += 1
    }
    
    while i < chars.count, chars[i].isNumber {
        result = result * 10 + chars[i].wholeNumberValue!
        i += 1
    }
    
    return sign * result
}
